### PR TITLE
nicely handle empty data for visualizations in SAM

### PR DESCRIPTION
**Pull request by moontrip**
This will address https://github.com/VEuPathDB/web-eda/issues/1631.

Following on from some changes in back end responses (see #836) we are revisiting the behaviour of floating plots when there is nothing to plot. A key background point here is that the SAM floating plot endpoints (e.g. histogram, lineplot etc) don't return `completeCasesAllVars` and `completeCasesAxesVars` for performance reasons and because missing data is not handled the same as in EDA. These two values are used to calculate `outputSize` which is fed to `PluginError` to produce an appropriate warning banner when it is zero.

In addition we have relaxed `io-ts` validation to allow nulls for binSlider and binSpec when there is an empty response from the back end (but no 204 or 500). This happens when there is a subset, but the plot variables are absent from that subset.

### old PR notes

The error message(s) indicates type errors of backend response. I think that null returns of binSlider/binSpec in Lineplot Viz means empty data return, thus relevant types are loose to accommodate null values. After this change, the error message is the same with what we could see at Scatterplot Viz

lineplot error message02

### CODE REVIEW

**Comment by bobular** [comment]
Maybe fix histogram too?

**Review by asizemore — APPROVED** [review]
Thanks for taking this @moontrip !

Works great! Tested with a different study.

I didn't quite understand one thing - why not change `BinSpec` and `BinSlider` to accommodate nulls?

**Comment by moontrip** [comment]
> Thanks for taking this @moontrip !
> 
> Works great! Tested with a different study.
> 
> I didn't quite understand one thing - why not change `BinSpec` and `BinSlider` to accommodate nulls?

Thank you for your review and test, @asizemore 👍 
I didn't work for typing Lineplot Viz initially so I don't know why they didn't accommodate nulls. My pure guess is that at the beginning of the implementation of the Lineplot Viz, we wanted to check whether the backend returned the response correctly even if data existed.

**Comment by moontrip** [comment]
@bobular: it would be great if you can review this PR. 
This PR will also address https://github.com/VEuPathDB/web-monorepo/issues/842. Refactored this PR to include Histogram Viz as well. Used the pre-existing message text at Histogram Viz.

- examples: Vectorbase Mega study

SAM histogram null04

SAM historgram null03

File: packages/libs/eda/src/lib/core/components/visualizations/implementations/HistogramVisualization.tsx [hunk 1]
```diff
@@ -1411,7 +1411,17 @@ export function histogramResponseToData(
...
+  // dataExist before filter() will have an array of booleans, e.g.,  [true, false, true,...]
+  // if data does not exists, the array consists of all false values
+  // filtering out true cases, then no data  = dataExist.length === 0
+  const dataExist = response.histogram.data
+    .map((item) => {
+      return item.value.length > 0;
+    })
+    .filter((element: boolean) => element);
+
+  if (response.histogram.data.length === 0 || dataExist.length === 0)
```
[the remark this answers is not in the document]
Comment (moontrip) [reply packages/libs/eda/src/lib/core/components/visualizations/implementations/HistogramVisualization.tsx]: Thank you for your suggestion, @bobular 👍 Complied!

File: packages/libs/eda/src/lib/core/components/visualizations/implementations/HistogramVisualization.tsx [hunk 2]
```diff
@@ -1411,7 +1411,7 @@ export function histogramResponseToData(
   overlayVariable?: Variable,
   facetVariable?: Variable
 ): HistogramDataWithCoverageStatistics {
-  if (response.histogram.data.length === 0)
+  if (!response.histogram.data.find((item) => item.value.length > 0))
     throw Error(`Expected one or more data series, but got zero`);
```
Comment (bobular) [inline packages/libs/eda/src/lib/core/components/visualizations/implementations/HistogramVisualization.tsx:1415]: This is an old error message that was never/rarely seen by users mainly because null values were causing `io-ts` to complain first.

I have removed this test entirely and found that nothing bad happens. However there is an empty plot with no explanation to the user. Which is bad. However, the `PluginError` component should handle this for us, but `outputSize` being passed to it is `undefined`.  We could either ensure that `outputSize` is zero before it is passed, or change the login in `PluginError` to this:

```diff
diff --git a/packages/libs/eda/src/lib/core/components/visualizations/PluginError.tsx b/packages/libs/eda/src/lib/core/components/visualizations/PluginError.tsx
index 51956105a3..6f8a1ea0b8 100644
--- a/packages/libs/eda/src/lib/core/components/visualizations/PluginError.tsx
+++ b/packages/libs/eda/src/lib/core/components/visualizations/PluginError.tsx
@@ -23,7 +23,7 @@ export default function PluginError({ error, customCases, outputSize }: Props) {
   const fallbackErrorMessage =
     error == null ? '' : error instanceof Error ? error.message : String(error);
 
-  const emptyCase = () => (outputSize === 0 ? emptyCaseMessage : undefined);
+  const emptyCase = () => (!outputSize ? emptyCaseMessage : undefined);
 
   const errorContent =
     (customCases ?? [])
```

I think I prefer the latter.

**Review by bobular — APPROVED** [review]
Please remove the test/throw for empty histogram data and modify `PluginError` as suggested.  Any questions, just shout!

File: packages/libs/eda/src/lib/core/components/visualizations/implementations/HistogramVisualization.tsx (hunk 2 above)
Comment (moontrip) [reply packages/libs/eda/src/lib/core/components/visualizations/implementations/HistogramVisualization.tsx:1415]: I realized that the latter did not work. It is because SAM's histogram always return outputSize = undefined. Thus, if we want to use PluginError, then we need to add an inline condition before passing to PluginError

File: packages/libs/eda/src/lib/core/components/visualizations/implementations/HistogramVisualization.tsx [hunk 3]
```diff
@@ -651,6 +652,18 @@ function HistogramViz(props: VisualizationProps<Options>) {
     ])
   );
 
+  // controls need the bin info from just one facet (not an empty one)
+  const checkData = isFaceted(data.value)
```
Comment (bobular) [inline packages/libs/eda/src/lib/core/components/visualizations/implementations/HistogramVisualization.tsx:656]: Need to get rid of `data0` below and replace it with `checkData`

**Comment by moontrip** [comment]
@bobular I have implemented the logic we discussed at Slack. It took more time than I expected to address empty plot issue. In one case, there is empty data response, not null case. In that case, Histogram Viz treated it as facetedData, so I needed to add a condition to handle it as not. It worked fine for me but it may also cause an edge case. 

With all various use cases that we cannot test all of them, honestly I am not quite sure if this is better than previous logic. But anyway this seems to work fine for SAM's Histogram VIz.

File: packages/libs/eda/src/lib/core/components/visualizations/implementations/HistogramVisualization.tsx (hunk 3 above)
Comment (moontrip) [reply packages/libs/eda/src/lib/core/components/visualizations/implementations/HistogramVisualization.tsx:656]: oh I see. will do that

Comment (moontrip) [reply packages/libs/eda/src/lib/core/components/visualizations/implementations/HistogramVisualization.tsx:656]: done!

**Comment by bobular** [comment]
Didn't do any special `isEmptyData` handling for the contingency table. After relaxing the io-ts constraints, it renders with zeroes and NaN% just fine, in my opinion

image

**Review by jernestmyers — APPROVED** [review]
I didn't test on a dev site, but the code looks okay to me. 💪

**Comment by asizemore** [comment]
tested boxplots and lineplots on mbio. Looks great!!

**Review by asizemore — APPROVED** [review]
I had a few questions but nothing serious!
